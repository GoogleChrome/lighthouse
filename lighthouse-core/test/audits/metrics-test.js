/**
 * @license Copyright 2018 The Lighthouse Authors. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
 */
'use strict';

import {jest} from '@jest/globals';

import MetricsAudit from '../../audits/metrics.js';
import TTIComputed from '../../computed/metrics/interactive.js';
import pwaTrace from '../fixtures/traces/progressive-app-m60.json';
import pwaDevtoolsLog from '../fixtures/traces/progressive-app-m60.devtools.log.json';
import lcpTrace from '../fixtures/traces/lcp-m78.json';
import lcpDevtoolsLog from '../fixtures/traces/lcp-m78.devtools.log.json';
import lcpAllFramesTrace from '../fixtures/traces/frame-metrics-m89.json';
import lcpAllFramesDevtoolsLog from '../fixtures/traces/frame-metrics-m89.devtools.log.json'; // eslint-disable-line max-len
import clsAllFramesTrace from '../fixtures/traces/frame-metrics-m90.json';
import clsAllFramesDevtoolsLog from '../fixtures/traces/frame-metrics-m90.devtools.log.json'; // eslint-disable-line max-len
import jumpyClsTrace from '../fixtures/traces/jumpy-cls-m90.json';
import jumpyClsDevtoolsLog from '../fixtures/traces/jumpy-cls-m90.devtoolslog.json';
import {getURLArtifactFromDevtoolsLog} from '../test-utils.js';

/* eslint-env jest */

describe('Performance: metrics', () => {
  it('evaluates valid input correctly', async () => {
    const URL = getURLArtifactFromDevtoolsLog(pwaDevtoolsLog);
    const artifacts = {
      URL,
      GatherContext: {gatherMode: 'navigation'},
      traces: {
        [MetricsAudit.DEFAULT_PASS]: pwaTrace,
      },
      devtoolsLogs: {
        [MetricsAudit.DEFAULT_PASS]: pwaDevtoolsLog,
      },
    };

    const context = {settings: {throttlingMethod: 'simulate'}, computedCache: new Map()};
    const result = await MetricsAudit.audit(artifacts, context);
    expect(result.details.items[0]).toMatchSnapshot();
  });

  it('evaluates valid input correctly (throttlingMethod=provided)', async () => {
    const URL = getURLArtifactFromDevtoolsLog(pwaDevtoolsLog);
    const artifacts = {
      URL,
      GatherContext: {gatherMode: 'navigation'},
      traces: {
        [MetricsAudit.DEFAULT_PASS]: pwaTrace,
      },
      devtoolsLogs: {
        [MetricsAudit.DEFAULT_PASS]: pwaDevtoolsLog,
      },
    };

    const context = {settings: {throttlingMethod: 'provided'}, computedCache: new Map()};
    const result = await MetricsAudit.audit(artifacts, context);
    expect(result.details.items[0]).toMatchSnapshot();
  });

  it('evaluates valid input (with lcp) correctly', async () => {
    const URL = getURLArtifactFromDevtoolsLog(lcpDevtoolsLog);
    const artifacts = {
      URL,
      GatherContext: {gatherMode: 'navigation'},
      traces: {
        [MetricsAudit.DEFAULT_PASS]: lcpTrace,
      },
      devtoolsLogs: {
        [MetricsAudit.DEFAULT_PASS]: lcpDevtoolsLog,
      },
    };

    const context = {settings: {throttlingMethod: 'simulate'}, computedCache: new Map()};
    const result = await MetricsAudit.audit(artifacts, context);
    expect(result.details.items[0]).toMatchSnapshot();
  });

  it('evaluates valid input (with lcp from all frames) correctly', async () => {
    const URL = getURLArtifactFromDevtoolsLog(lcpAllFramesDevtoolsLog);
    const artifacts = {
      URL,
      GatherContext: {gatherMode: 'navigation'},
      traces: {
        [MetricsAudit.DEFAULT_PASS]: lcpAllFramesTrace,
      },
      devtoolsLogs: {
        [MetricsAudit.DEFAULT_PASS]: lcpAllFramesDevtoolsLog,
      },
    };

    const context = {settings: {throttlingMethod: 'provided'}, computedCache: new Map()};
    const result = await MetricsAudit.audit(artifacts, context);
    expect(result.details.items[0]).toMatchSnapshot();
  });

  it('leaves CLS undefined in an old trace without weighted scores', async () => {
    const artifacts = {
      GatherContext: {gatherMode: 'navigation'},
      traces: {
        [MetricsAudit.DEFAULT_PASS]: lcpAllFramesTrace,
      },
      devtoolsLogs: {
        [MetricsAudit.DEFAULT_PASS]: lcpAllFramesDevtoolsLog,
      },
    };

    const context = {settings: {throttlingMethod: 'simulate'}, computedCache: new Map()};
    const {details} = await MetricsAudit.audit(artifacts, context);
    expect(details.items[0]).toMatchObject({
      cumulativeLayoutShift: undefined,
      cumulativeLayoutShiftMainFrame: undefined,
      totalCumulativeLayoutShift: undefined,
      observedCumulativeLayoutShift: undefined,
      observedCumulativeLayoutShiftMainFrame: undefined,
      observedTotalCumulativeLayoutShift: undefined,
    });
  });

  it('evaluates new CLS correctly across all frames', async () => {
    const URL = getURLArtifactFromDevtoolsLog(clsAllFramesDevtoolsLog);
    const artifacts = {
      URL,
      GatherContext: {gatherMode: 'navigation'},
      traces: {
        [MetricsAudit.DEFAULT_PASS]: clsAllFramesTrace,
      },
      devtoolsLogs: {
        [MetricsAudit.DEFAULT_PASS]: clsAllFramesDevtoolsLog,
      },
    };

    const context = {settings: {throttlingMethod: 'provided'}, computedCache: new Map()};
    const {details} = await MetricsAudit.audit(artifacts, context);

    // Only a single main-frame shift event, so mfCls and oldCls are equal.
    expect(details.items[0]).toMatchObject({
      cumulativeLayoutShift: expect.toBeApproximately(0.026463, 6),
      cumulativeLayoutShiftMainFrame: expect.toBeApproximately(0.001166, 6),
      totalCumulativeLayoutShift: expect.toBeApproximately(0.001166, 6),

      observedCumulativeLayoutShift: expect.toBeApproximately(0.026463, 6),
      observedCumulativeLayoutShiftMainFrame: expect.toBeApproximately(0.001166, 6),
      observedTotalCumulativeLayoutShift: expect.toBeApproximately(0.001166, 6),
    });
  });

  it('does not fail the entire audit when TTI errors', async () => {
    const URL = getURLArtifactFromDevtoolsLog(pwaDevtoolsLog);
    const artifacts = {
      URL,
      GatherContext: {gatherMode: 'navigation'},
      traces: {
        [MetricsAudit.DEFAULT_PASS]: pwaTrace,
      },
      devtoolsLogs: {
        [MetricsAudit.DEFAULT_PASS]: pwaDevtoolsLog,
      },
    };

    const mockTTIFn = jest.spyOn(TTIComputed, 'request');
    mockTTIFn.mockRejectedValueOnce(new Error('TTI failed'));
    const context = {settings: {throttlingMethod: 'simulate'}, computedCache: new Map()};
    const result = await MetricsAudit.audit(artifacts, context);
    expect(result.details.items[0].interactive).toEqual(undefined);
  });

  it('evaluates CLS correctly', async () => {
    const URL = getURLArtifactFromDevtoolsLog(jumpyClsDevtoolsLog);
    const artifacts = {
      URL,
      GatherContext: {gatherMode: 'navigation'},
      traces: {
        [MetricsAudit.DEFAULT_PASS]: jumpyClsTrace,
      },
      devtoolsLogs: {
        [MetricsAudit.DEFAULT_PASS]: jumpyClsDevtoolsLog,
      },
    };

    const context = {settings: {throttlingMethod: 'simulate'}, computedCache: new Map()};
    const {details} = await MetricsAudit.audit(artifacts, context);
    expect(details.items[0]).toMatchObject({
      cumulativeLayoutShift: expect.toBeApproximately(2.268816, 6),
      cumulativeLayoutShiftMainFrame: expect.toBeApproximately(2.268816, 6),
      totalCumulativeLayoutShift: expect.toBeApproximately(4.809794, 6),

      observedCumulativeLayoutShift: expect.toBeApproximately(2.268816, 6),
      observedCumulativeLayoutShiftMainFrame: expect.toBeApproximately(2.268816, 6),
      observedTotalCumulativeLayoutShift: expect.toBeApproximately(4.809794, 6),
    });
  });
});
