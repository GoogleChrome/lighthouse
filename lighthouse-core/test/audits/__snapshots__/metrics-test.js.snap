// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Performance: metrics evaluates valid input (with lcp from all frames) correctly 1`] = `
Object {
  "cumulativeLayoutShift": 0.0011656245471340055,
  "cumulativeLayoutShiftAllFrames": 0.5436596106821069,
  "estimatedInputLatency": 16,
  "estimatedInputLatencyTs": undefined,
  "firstCPUIdle": 688,
  "firstCPUIdleTs": 46134430620,
  "firstContentfulPaint": 688,
  "firstContentfulPaintTs": 46134430620,
  "firstMeaningfulPaint": 688,
  "firstMeaningfulPaintTs": 46134430620,
  "interactive": 688,
  "interactiveTs": 46134430620,
  "largestContentfulPaint": 688,
  "largestContentfulPaintAllFrames": 5948,
  "largestContentfulPaintAllFramesTs": 46139690898,
  "largestContentfulPaintTs": 46134430620,
  "maxPotentialFID": 16,
  "observedCumulativeLayoutShift": 0.0011656245471340055,
  "observedCumulativeLayoutShiftAllFrames": 0.5436596106821069,
  "observedDomContentLoaded": 617,
  "observedDomContentLoadedTs": 46134359407,
  "observedFirstContentfulPaint": 688,
  "observedFirstContentfulPaintTs": 46134430620,
  "observedFirstMeaningfulPaint": 688,
  "observedFirstMeaningfulPaintTs": 46134430620,
  "observedFirstPaint": 688,
  "observedFirstPaintTs": 46134430620,
  "observedFirstVisualChange": 679,
  "observedFirstVisualChangeTs": 46134421490,
  "observedLargestContentfulPaint": 688,
  "observedLargestContentfulPaintAllFrames": 5948,
  "observedLargestContentfulPaintAllFramesTs": 46139690898,
  "observedLargestContentfulPaintTs": 46134430620,
  "observedLastVisualChange": 5967,
  "observedLastVisualChangeTs": 46139709490,
  "observedLoad": 706,
  "observedLoadTs": 46134448526,
  "observedNavigationStart": 0,
  "observedNavigationStartTs": 46133742490,
  "observedSpeedIndex": 1370,
  "observedSpeedIndexTs": 46135112850,
  "observedTimeOrigin": 0,
  "observedTimeOriginTs": 46133742490,
  "observedTraceEnd": 6019,
  "observedTraceEndTs": 46139761594,
  "speedIndex": 1370,
  "speedIndexTs": 46135112490,
  "totalBlockingTime": 0,
}
`;

exports[`Performance: metrics evaluates valid input (with lcp) correctly 1`] = `
Object {
  "cumulativeLayoutShift": 0,
  "cumulativeLayoutShiftAllFrames": 0,
  "estimatedInputLatency": 543,
  "estimatedInputLatencyTs": undefined,
  "firstCPUIdle": 3677,
  "firstCPUIdleTs": undefined,
  "firstContentfulPaint": 2289,
  "firstContentfulPaintTs": undefined,
  "firstMeaningfulPaint": 2758,
  "firstMeaningfulPaintTs": undefined,
  "interactive": 4462,
  "interactiveTs": undefined,
  "largestContentfulPaint": 2758,
  "largestContentfulPaintAllFrames": undefined,
  "largestContentfulPaintAllFramesTs": undefined,
  "largestContentfulPaintTs": undefined,
  "maxPotentialFID": 1336,
  "observedCumulativeLayoutShift": 0,
  "observedCumulativeLayoutShiftAllFrames": 0,
  "observedDomContentLoaded": 1513,
  "observedDomContentLoadedTs": 713038536140,
  "observedFirstContentfulPaint": 1122,
  "observedFirstContentfulPaintTs": 713038144775,
  "observedFirstMeaningfulPaint": 1122,
  "observedFirstMeaningfulPaintTs": 713038144775,
  "observedFirstPaint": 1122,
  "observedFirstPaintTs": 713038144775,
  "observedFirstVisualChange": 1105,
  "observedFirstVisualChangeTs": 713038128064,
  "observedLargestContentfulPaint": 1122,
  "observedLargestContentfulPaintAllFrames": undefined,
  "observedLargestContentfulPaintAllFramesTs": undefined,
  "observedLargestContentfulPaintTs": 713038144775,
  "observedLastVisualChange": 1722,
  "observedLastVisualChangeTs": 713038745064,
  "observedLoad": 2159,
  "observedLoadTs": 713039182071,
  "observedNavigationStart": 0,
  "observedNavigationStartTs": 713037023064,
  "observedSpeedIndex": 1393,
  "observedSpeedIndexTs": 713038416494,
  "observedTimeOrigin": 0,
  "observedTimeOriginTs": 713037023064,
  "observedTraceEnd": 7416,
  "observedTraceEndTs": 713044439102,
  "speedIndex": 3681,
  "speedIndexTs": undefined,
  "totalBlockingTime": 1167,
}
`;

exports[`Performance: metrics evaluates valid input correctly (throttlingMethod=provided) 1`] = `
Object {
  "cumulativeLayoutShift": 0,
  "cumulativeLayoutShiftAllFrames": 0,
  "estimatedInputLatency": 17,
  "estimatedInputLatencyTs": undefined,
  "firstCPUIdle": 1582,
  "firstCPUIdleTs": 225415754204,
  "firstContentfulPaint": 499,
  "firstContentfulPaintTs": 225414670885,
  "firstMeaningfulPaint": 783,
  "firstMeaningfulPaintTs": 225414955343,
  "interactive": 1582,
  "interactiveTs": 225415754204,
  "largestContentfulPaint": undefined,
  "largestContentfulPaintAllFrames": undefined,
  "largestContentfulPaintAllFramesTs": undefined,
  "largestContentfulPaintTs": undefined,
  "maxPotentialFID": 198,
  "observedCumulativeLayoutShift": 0,
  "observedCumulativeLayoutShiftAllFrames": 0,
  "observedDomContentLoaded": 560,
  "observedDomContentLoadedTs": 225414732309,
  "observedFirstContentfulPaint": 499,
  "observedFirstContentfulPaintTs": 225414670885,
  "observedFirstMeaningfulPaint": 783,
  "observedFirstMeaningfulPaintTs": 225414955343,
  "observedFirstPaint": 499,
  "observedFirstPaintTs": 225414670868,
  "observedFirstVisualChange": 520,
  "observedFirstVisualChangeTs": 225414692015,
  "observedLargestContentfulPaint": undefined,
  "observedLargestContentfulPaintAllFrames": undefined,
  "observedLargestContentfulPaintAllFramesTs": undefined,
  "observedLargestContentfulPaintTs": undefined,
  "observedLastVisualChange": 818,
  "observedLastVisualChangeTs": 225414990015,
  "observedLoad": 2199,
  "observedLoadTs": 225416370913,
  "observedNavigationStart": 0,
  "observedNavigationStartTs": 225414172015,
  "observedSpeedIndex": 605,
  "observedSpeedIndexTs": 225414776724,
  "observedTimeOrigin": 0,
  "observedTimeOriginTs": 225414172015,
  "observedTraceEnd": 12540,
  "observedTraceEndTs": 225426711887,
  "speedIndex": 605,
  "speedIndexTs": 225414777015,
  "totalBlockingTime": 48,
}
`;

exports[`Performance: metrics evaluates valid input correctly 1`] = `
Object {
  "cumulativeLayoutShift": 0,
  "cumulativeLayoutShiftAllFrames": 0,
  "estimatedInputLatency": 78,
  "estimatedInputLatencyTs": undefined,
  "firstCPUIdle": 3351,
  "firstCPUIdleTs": undefined,
  "firstContentfulPaint": 1337,
  "firstContentfulPaintTs": undefined,
  "firstMeaningfulPaint": 1553,
  "firstMeaningfulPaintTs": undefined,
  "interactive": 3427,
  "interactiveTs": undefined,
  "largestContentfulPaint": undefined,
  "largestContentfulPaintAllFrames": undefined,
  "largestContentfulPaintAllFramesTs": undefined,
  "largestContentfulPaintTs": undefined,
  "maxPotentialFID": 396,
  "observedCumulativeLayoutShift": 0,
  "observedCumulativeLayoutShiftAllFrames": 0,
  "observedDomContentLoaded": 560,
  "observedDomContentLoadedTs": 225414732309,
  "observedFirstContentfulPaint": 499,
  "observedFirstContentfulPaintTs": 225414670885,
  "observedFirstMeaningfulPaint": 783,
  "observedFirstMeaningfulPaintTs": 225414955343,
  "observedFirstPaint": 499,
  "observedFirstPaintTs": 225414670868,
  "observedFirstVisualChange": 520,
  "observedFirstVisualChangeTs": 225414692015,
  "observedLargestContentfulPaint": undefined,
  "observedLargestContentfulPaintAllFrames": undefined,
  "observedLargestContentfulPaintAllFramesTs": undefined,
  "observedLargestContentfulPaintTs": undefined,
  "observedLastVisualChange": 818,
  "observedLastVisualChangeTs": 225414990015,
  "observedLoad": 2199,
  "observedLoadTs": 225416370913,
  "observedNavigationStart": 0,
  "observedNavigationStartTs": 225414172015,
  "observedSpeedIndex": 605,
  "observedSpeedIndexTs": 225414776724,
  "observedTimeOrigin": 0,
  "observedTimeOriginTs": 225414172015,
  "observedTraceEnd": 12540,
  "observedTraceEndTs": 225426711887,
  "speedIndex": 1676,
  "speedIndexTs": undefined,
  "totalBlockingTime": 726,
}
`;
